ENH: better dtype inference when doing DataFrame reductions

topper-123 opened the pull request:
- [x] closes #40669, #42895 & #14162

Supercedes #52707 and possibly #52261, depending on reception.

This PR improves on #52707 by not attempting to infer the combined dtypes from the original dtypes and builds heavily on the ideas in #52261, but avoids the overloading of `kwargs` in the extensionarray functions/methods in that PR. Instead this  keeps all  the work in `ExtensionArray._reduce` by adding an explicit `keepdims` param to that method signature. This makes this implementation simpler than #52261 IMO.

This PR is not finished, and would I appreciate feedback about the direction here compared to #52261 before I proceed. This  works currently correctly for masked arrays & ArrowArrays (AFAIKS), but still need the other extensionArray subclasses (datetimelike arrays & Categorical). I've written some tests, but need to write more + write the docs.

CC: @jbrockmendel & @rhshadrach

topper-123 committed:
ENH: better dtype inference when doing DataFrame reductions

topper-123 committed:
precommit issues

jbrockmendel commented on pandas/core/arrays/masked.py:
```diff
@@ -1097,6 +1108,30 @@ def _wrap_reduction_result(self, name: str, result, skipna, **kwargs):
+    def _wrap_na_result(self, name):
+        mask = np.ones(1, dtype=bool)
+
+        if is_float_dtype(self.dtype):
```
any chance of avoiding these?  weve been moving away from them for perf reasons

topper-123 replied:
Ok.

jbrockmendel commented:
The reason for the gymnastics/checking for keepdims in #52261 is bc 3rd party EAs won't have implemented keepdims, so will need a deprecation cycle to catch up (though in general I'd be OK with going less far out of our way with this kind of thing)

jbrockmendel commented on pandas/core/arrays/base.py:
```diff
@@ -1403,7 +1403,9 @@ def _accumulate(
 
-    def _reduce(self, name: str, *, skipna: bool = True, **kwargs):
+    def _reduce(
+        self, name: str, *, skipna: bool = True, keepdims: bool = False, **kwargs
```
probably need keepdims in the docstring

topper-123 replied:
Added.

jbrockmendel commented on pandas/core/arrays/base.py:
```diff
@@ -1433,7 +1435,14 @@ def _reduce(self, name: str, *, skipna: bool = True, **kwargs):
+        if keepdims:
+            # if subclasses want to avoid wrapping in np.array, do:
+            # super()._reduce(..., keepdims=False) and wrap that.
+            return np.array([[result]])
```
should we warn/deprecate here?

topper-123 replied:
This reminds me that using `*args` & `**kwargs` in the method signature makes changing APIs difficult, in addition to them making typing more difficult. I really think they should be avoided whenever possible.

Is there a good/clean way to add a deprecation? I assume if any user has been using a `keepdims` parameter in their code, it's likely been a bool, so using type checking for deprecation inside `_reduce` would fail in those cases. We could check `kwargs` in `blk_func` in `frame.py` for key `"keepdims"`. That depends on us not calling `_reduce(keepdims)` in any other location (we probably won't) or remembering to add deprecation texts there also. Any better idea?

topper-123 committed:
fix failures

topper-123 committed:
fix failures

topper-123 committed:
mypy + some docs

topper-123 committed:
doc linting linting

topper-123 committed:
refactor to use _reduce_with_wrap

topper-123 commented:
I've refactored to use `_reduce_with_wrap` method instead of a keyword. This avoids the deprercation issue raised by @jbrockmendel.

topper-123 committed:
docstring linting

topper-123 committed:
pyarrow failure + linting

topper-123 committed:
pyarrow failure + linting

topper-123 committed:
linting

topper-123 committed:
doc stuff

topper-123 commented:
Actually it is backward compatible: If they don't implement this, their scalar just gets wrapped in a numpy array, so will work unchanged. So it's backward compatible, they just don't get this improvement without implementing `_reduce_with_wrap` in their arrays.

rhshadrach commented on pandas/core/arrays/masked.py:
```diff
@@ -1097,6 +1103,30 @@ def _wrap_reduction_result(self, name: str, result, skipna, **kwargs):
+    def _wrap_na_result(self, name):
+        mask = np.ones(1, dtype=bool)
+
+        if self.dtype.kind == "f":
+            np_dtype = "float64"
+        elif name in ["mean", "median", "var", "std", "skew"]:
+            np_dtype = "float64"
+        elif self.dtype.kind == "i":
+            np_dtype = "int64"
+        elif self.dtype.kind == "u":
+            np_dtype = "uint64"
+        else:
+            raise TypeError(self.dtype)
```
Should  corr, sem, and kurt be part of the first elif? And is this not hit by e.g. a sum of Boolean?

In general I'm not a fan of encoding op-specific logic like this. It can lead to bugs if behavior is changed in one place but not the other.

topper-123 replied:
Yeah, good points, I'll have to look into this again to see if I can do it differently.

rhshadrach replied:
Wow - nice! So this wasn't necessary?

topper-123 replied:
yeah It was needed in the first draft, then I changed the code making it unnecessary.

There is still some things missing in this PR. I'll update today.

rhshadrach commented on pandas/tests/extension/base/dim2.py:
```diff
@@ -197,6 +197,8 @@ def test_reductions_2d_axis0(self, data, method):
         if method in ["std", "var"]:
             # pass ddof=0 so we get all-zero std instead of all-NA std
             kwargs["ddof"] = 0
+        elif method in ["prod", "sum"]:
+            kwargs["min_count"] = 1
```
Why does this need to change?

topper-123 replied:
The input array ('arr2d') contains NA values. Previously we were adding scalars, so e.g. 'sum([pd.NA])' would give `pd.NA`, even with `min_count=0`.

In the new version, pandas will be able to correctly give result of 0, if `min_count=0` while correctly giving result `pd.NA`, with `min_count=1`.

So this could be considered a bug fix.

topper-123 replied:
There's isn't currently for the 2-dim case, only the 1-dim case. I can add it, but that will have to be tomorrow.

topper-123 replied:
I've updated this.

topper-123 committed:
linting fixes

topper-123 committed:
fix fix doc string

topper-123 committed:
remove _wrap_na_result

topper-123 committed:
BUG: Adding a columns to a Frame with RangeIndex columns using a non-scalar key (#52877)

topper-123 committed:
doc string example

phofl committed:
DOC: Update whatsnew (#52882)

topper-123 committed:
pyarrow + categorical

topper-123 commented:
I've updated, so it now works correctly with pyarrow and categorical dtypes. Datetimelikes still missing, likewise docs and maybe some aditional tests.

topper-123 committed:
silence bugs

topper-123 committed:
silence errors

topper-123 committed:
silence errors II

topper-123 committed:
fix errors III

topper-123 committed:
various fixups

topper-123 committed:
various fixups

topper-123 committed:
delay fixing windows and 32bit failures

phofl committed:
CI: Change development python version to 3.10 (#51133)

* CI: Change development python version to 3.10

* Update checks

* Remove strict

* Remove strict

* Fixes

* Add dt

* Switch python to 3.9

* Remove

* Fix

* Try attribute

* Adjust

* Fix mypy

* Try fixing doc build

* Fix mypy

* Fix stubtest

* Remove workflow file

* Rename back

* Update

* Rename

* Rename

* Change python version

* Remove

* Fix doc errors

* Remove pypy

* Update ci/deps/actions-pypy-39.yaml

Co-authored-by: Thomas Li <47963215+lithomas1@users.noreply.github.com>

* Revert pypy removal

* Remove again

* Fix

* Change to 3.9

* Address

---------

Co-authored-by: Thomas Li <47963215+lithomas1@users.noreply.github.com>

topper-123 committed:
update

topper-123 committed:
update

topper-123 committed:
add docs

jbrockmendel commented on pandas/core/arrays/arrow/array.py:
```diff
@@ -1310,6 +1312,12 @@ def pyarrow_meth(data, skip_nulls, **kwargs):
             result = result[0]
+
+        if keepdims:
+            # TODO: is there a way to do this without .as_py()
```
@jorisvandenbossche gentle ping, i don't want to move forward here without your OK

topper-123 replied:
This just convert a scalar, so perforemancewise the effect from this is limited. Would be worse if it was a whole array being converted.

jorisvandenbossche replied:
Yes, using `as_py()` is the sensible thing to do here to convert a scalar to a len-1 array (until pyarrow supports it's own scalars in `pa.array([..])`)

(we have internal functionality in C++ to convert a scalar to a len-1 array, but not sure that is worth exposing in pyarrow)

topper-123 committed:
fix windows tests

topper-123 committed:
fix windows tests

topper-123 committed:
remove guards for 32bit linux

topper-123 committed:
add bool tests + fix 32-bit failures

topper-123 committed:
fix pre-commit failures

topper-123 committed:
fix mypy failures

topper-123 committed:
rename _reduce_with -> _reduce_and_wrap

rhshadrach commented on pandas/tests/extension/test_boolean.py:
```diff
@@ -367,6 +372,30 @@ def check_reduce(self, s, op_name, skipna):
 
+    def check_reduce_and_wrap(self, ser: pd.Series, op_name: str, skipna: bool):
+        if op_name in ["count", "kurt", "sem"]:
+            pytest.skip(f"{op_name} not an array method")
```
Same

topper-123 replied:
Added.

topper-123 commented:
After the various perf PRs ( #52998, #53013 & #53040), I now get unchanged perf for axis=0 when calling `asv continuous -f 1.1 upstream/main HEAD -b stat_ops.FrameOps`:

```
       before           after         ratio
     [4f14b456]       [5760ee38]
     <master>         <reduction_dtypes_II>
+      3.90±0.02s       12.7±0.01s     3.25  stat_ops.FrameOps.time_op('median', 'Int64', 1)
+      4.22±0.01s       6.68±0.01s     1.58  stat_ops.FrameOps.time_op('kurt', 'Int64', 1)
+         3.96±0s       5.86±0.01s     1.48  stat_ops.FrameOps.time_op('skew', 'Int64', 1)
+      2.94±0.01s       3.82±0.01s     1.30  stat_ops.FrameOps.time_op('mean', 'Int64', 1)
+         2.34±0s       3.03±0.01s     1.29  stat_ops.FrameOps.time_op('sum', 'Int64', 1)
+      2.35±0.01s       3.01±0.01s     1.28  stat_ops.FrameOps.time_op('prod', 'Int64', 1)
+      3.70±0.01s       4.58±0.01s     1.24  stat_ops.FrameOps.time_op('std', 'Int64', 1)
+      3.63±0.01s       4.50±0.01s     1.24  stat_ops.FrameOps.time_op('var', 'Int64', 1)
-     3.46±0.03ms      3.08±0.02ms     0.89  stat_ops.FrameOps.time_op('median', 'Int64', 0)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

That means that `axis=0` is good now, while for `axis=1` is slower because it's in effect the same as doing `df.T.sum(axis=0)` on a (4, 100_000,)-shaped DataFrame, where the transpose is very slow (but can be implroved by #52836 or #52083) + there is a lot of work to extract masked  data from 100_000 ExtensionBlocks. So I think this is as good as it gets for `axis=1`, unless we can avoid transposing the dataframe, which will speed things up tremendously, but that's unrelated to this PR.

topper-123 commented on pandas/core/arrays/masked.py:
```diff
@@ -108,6 +108,13 @@ class BaseMaskedArray(OpsMixin, ExtensionArray):
     _falsey_value = Scalar  # bool(_falsey_value) = False
 
+    @classmethod
+    def _simple_new(cls, values: np.ndarray, mask: npt.NDArray[np.bool_]) -> Self:
+        result = BaseMaskedArray.__new__(cls)
+        result._data = values
+        result._mask = mask
+        return result
```
This is from #53013 (I built this on top of that to see what performance improvements I get), so can be handled independently in #53013.

and yes, this is to avoid the checks, when not needed. For this PR i need `BaseMaskedArray.reshape` to be fast and this helps with that (and many other cases).

topper-123 commented on pandas/tests/extension/test_arrow.py:
```diff
@@ -507,6 +507,38 @@ def test_reduce_series(self, data, all_numeric_reductions, skipna, request):
 
+    def check_reduce_and_wrap(self, ser, op_name, skipna):
+        if op_name in ["count", "kurt", "sem", "skew"]:
+            pytest.skip(f"{op_name} not an array method")
```
Added.

topper-123 committed:
assert missing attributes

topper-123 commented on pandas/tests/extension/test_boolean.py:
```diff
@@ -367,6 +372,30 @@ def check_reduce(self, s, op_name, skipna):
+            cmp_dtype = "boolean"
+        elif op_name in ["sum", "prod"]:
+            is_windows_or_32bit = is_platform_windows() or not IS64
+            cmp_dtype = "Int32" if skipna and is_windows_or_32bit else "Int64"
```
I've had a big struggle to get a handle on how the dtype end up changing on different systems, so this is definitely not deliberate on my part, just trying to get the tests to pass, then it will be easier to fix up whatever needs to be fixed.

But I think you are probably right, I can look into it tomorrow. Having said that, I don't think I've changed anything in this regard, so if this is a bug, this just uncovers an existing bug and it shouldn't be not something that has been introduced in this PR.

topper-123 replied:
I've fixed this, so skipna doesn't affect the dtype.

Also: "I don't think I've changed anything in this regard." are always famous last words. My PR was the cause of this, so good catch.

topper-123 committed:
add tests for min_count=0

topper-123 committed:
reduction dtypes on windows and 32bit systems

topper-123 committed:
PERF:median with axis=1

topper-123 commented:
I've improved the `median` with axis=1 case, so it now falls in line with the other axis=1 reductions

```
       before           after         ratio
     [4f14b456]       [122488b1]
     <master>         <reduction_dtypes_II>
+      4.25±0.01s       6.68±0.02s     1.57  stat_ops.FrameOps.time_op('kurt', 'Int64', 1)
+      4.64±0.05s          6.42±0s     1.38  stat_ops.FrameOps.time_op('sem', 'Int64', 1)
+         2.35±0s       3.04±0.01s     1.29  stat_ops.FrameOps.time_op('sum', 'Int64', 1)
+      2.96±0.01s          3.81±0s     1.29  stat_ops.FrameOps.time_op('mean', 'Int64', 1)
+      3.60±0.04s          4.53±0s     1.26  stat_ops.FrameOps.time_op('var', 'Int64', 1)
+      3.70±0.01s          4.65±0s     1.26  stat_ops.FrameOps.time_op('std', 'Int64', 1)
+      3.92±0.01s       4.85±0.02s     1.24  stat_ops.FrameOps.time_op('median', 'Int64', 1)
+     1.36±0.05ms      1.50±0.07ms     1.11  stat_ops.FrameOps.time_op('kurt', 'Int64', 0)
-     3.45±0.07ms      3.02±0.04ms     0.88  stat_ops.FrameOps.time_op('median', 'Int64', 0)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE DECREASED.
```

Generally, in the `axis=1` cases in the tests, we now have to concatenating the reduction result for the 100_000 masked arrays.  Before we only had to concatenate 100_000 numpy arrays, which was inherently faster. OTOH doing concatenation over 100_000 masked arrays is always slow, so can't become much faster without a significant rewrite which will be out of scope here.

topper-123 committed:
median with axis=1 fix

jbrockmendel commented on pandas/core/arrays/base.py:
```diff
@@ -1434,6 +1441,28 @@ def _reduce(self, name: str, *, skipna: bool = True, **kwargs):
+        Length: 1, dtype: Int64
+        """
+        result = self._reduce(name, skipna=skipna, **kwargs)
+        return np.array([[result]])
```
this looks like a 2D result.  Should it be 1D?

topper-123 replied:
This just moved the existing behavior from `pandas.core.internals.blocks.Block.reduce` to here. Having said that, I think you are right, 1D makes more sense here and it passes all tests also, so I've changed it.

jbrockmendel commented on pandas/core/arrays/masked.py:
```diff
@@ -1098,25 +1108,62 @@ def _reduce(self, name: str, *, skipna: bool = True, **kwargs):
+            return result
         return result
 
+    def _wrap_na_result(self, *, name, axis):
```
this feels like it got a lot more complicated than the first attempt at "keepdims".  does this just address more corner cases the first attempt missed?

topper-123 replied:
Yeah, this is not great, and I've been tinkered quite a lot with other implementations. I think/hope I will have a simpler solution tonight.

topper-123 replied:
Ok, I looked into this and unless I do a relatively major refactoring, it looks like I just move complexity around by changing this. So without a bigger rework of `arrays_algos`, I don't see any clearly better method. 

Suggestions/ideas that show otherwise welcome, of course.

topper-123 replied:
to add, the underlying issue is that the functions in `masked_reductions.py` only return scalar `NA` values. This means that we don't get the type information when doing reductions that return `NA`, but have to infer it and the inferring is complex because it depends on calling method, dtype of calling data and OS platform. 

The solution would be to return proper 2D results from `masked_reductions.py` functions, e.g. return a tuple instead of scalar, e.g.
1.  return (scalar_value, none), when returning scalar
2. return (value_array, mask_array) when returning a masked array

and the wrap the result in an array  in `BaseMaskedArray._wrap_reduction_result`, when it should be a masked array.

However, I 'm not sure if that's the direction we want to pursue because unless we want to support 2d masked arrays, the current solution could still be simpler than building out 2d reduction support.

jorisvandenbossche replied:
Longer term, that indeed sounds as a good alternative, even for the current 1D-reshaped-as-2D case. One potential disadvantage is that you still need to do some actual calculation on the data, unnecessarily, to get the `value_array` return value (so we can use that to determine the result dtype). Calculating the actual value is "unnecessary" if you know the result will be masked. Of course it avoids hardcoding result dtypes. But for the case of skipna=False with a large array, that might introduce a slowdown?

topper-123 replied:
We only "know" because of the hard-coding here, for example `Series([1, 2, pd.NA], dtype="int8").sum()` is `Int32` on some systems and `Int64` on others. So to avoid the hardcoding we will have to do a calculation, so it's a trade-off (or we could hard-code the dtype when we're 100 % sure we get a NA, but that'll be about the same complexity as now + the extra code for doing real 2d ops). I think we should only do the 2d ops if we want it independent of this issue here. 

IMO that's a possible future PR, if we choose to go in that direction.

jorisvandenbossche replied:
> We only "know" because of the hard-coding here

In the end, also numpy kind of "hard codes" this, but just in the implementation itself. And we get the difficulty of combining the algo of numpy with our own logic, where we also want to know this information. So I don't think it's necessarily "wrong" to hard code this on our side as well (i.e. essentially keep it as you are doing in this PR, also long term (at least as long as the arrays are only 1D)).

Sidenote, from seeing the additional complexity for 32bit systems in deciding the result dtype, I do wonder if we actually want to get rid of that system-dependent behaviour? (we also don't follow numpy's behaviour in the constructors, but always default to int64) 
Although given that we still call numpy for the actual operation, it probably doesn't reduce complexity to ensure this dtype guarantee (we would need to move the 32bit check to the implementation of the masked sum)

jbrockmendel commented on pandas/core/internals/blocks.py:
```diff
@@ -339,7 +339,9 @@ def reduce(self, func) -> list[Block]:
             # TODO(EA2D): special case not needed with 2D EAs
-            res_values = np.array([[result]])
+            if not isinstance(result, (np.ndarray, ExtensionArray)):
+                result = np.array([[result]])
```
shouldn't this now be unreachable?

topper-123 replied:
Actually yes. I will update.

topper-123 replied:
I've updated.

jbrockmendel commented:
id expect this to be very close to perf-neutral.  where is the slowdown coming from?

topper-123 committed:
streamline Block.reduce

topper-123 committed:
fix comments

topper-123 commented on doc/source/whatsnew/v2.1.0.rst:
```diff
@@ -14,10 +14,43 @@ including other versions of pandas.
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In previous versions of pandas, the results of DataFrame reductions
+(:meth:`DataFrameG.sum` :meth:`DataFrame.mean` etc.) has numpy dtypes even when the DataFrames
```
> id expect this to be very close to perf-neutral. where is the slowdown coming from?

It is performance neutral when the dataframe is not-wide, but when the dataframe gets very wide, there will be a certain slowdown, because pandas will have to create 1 new masked array per column and then join those masked arrays. Joining masked arrays is slower than joining the same number of ndarrays, because the masked array have two ndarrays internally + the masked arrays are written in python while the ndarrays are written in C, which gives some overhead for masked arrays.

Having said that, note that the ASV slowdowns that I show is about  joining 100.000 columns, i.e quite a lot. I think the performance is actually ok, taking the above explanation into account.

topper-123 replied:
fixed.

topper-123 commented:
> id expect this to be very close to perf-neutral. where is the slowdown coming from?

It is performance neutral when the dataframe is not-wide, but when the dataframe gets very wide, there will be a certain slowdown, because pandas will have to create 1 new masked array per column and then join those masked arrays. Joining masked arrays is slower than joining the same number of ndarrays, because the masked array have two ndarrays internally + the masked arrays are written in python while the ndarrays are written in C, which gives some overhead for masked arrays.

Having said that, note that the ASV slowdowns that I show is about  joining 100.000 columns, i.e quite a lot. I think the performance is actually ok, taking the above explanation into account.

glemaitre committed:
FIX preserve dtype with datetime columns of different resolution when merging (#53213)

topper-123 commented:
Ping... The slowdown is for very wide arrays, so I don't think this is a practical issue, and the dtype inference when doing reduction is a very nice UX to get into pandas IMO.

Charlie-XIAO committed:
BUG Merge not behaving correctly when having `MultiIndex` with a single level (#53215)

* fix merge when MultiIndex with single level

* resolved conversations

* fixed code style

jorisvandenbossche committed:
BUG: preserve dtype for right/outer merge of datetime with different resolutions (#53233)

rhshadrach commented:
Do we know/suspect the `keepdims` version of this would experience the same slowdown?

topper-123 commented:
> Do we know/suspect the `keepdims` version of this would experience the same slowdown?

This PR does use `keepdims=True`, so I don't understand the question. Do you mean #52261?

Generally about performance here, the slower test is about aggregating 100.000 columns each being of length 2 and rejoining the result values into a series. In this scenario each aggregation will always be lightning fast, because each array is of length just 2. If the size of the data was different, the aggregation calculation cost per column would be much higher and that would dwarf the slowdown we see here. So the issue is specific to wide and short dataframes (which are rare) +  when aggregating over axis=1 (much rarer than aggregation over axis=0, but is already slow from the need to transform) and in return we get better dtype handling for masked arrays (which is very useful, so we solve a UX painpoint from using masked arrays by doing this).

rhshadrach commented:
> This PR does use `keepdims=True`, so I don't understand the question. Do you mean #52261?

Yes.

> Generally about performance here...

I think as-is this change is worth the performance impact we're seeing. I also agree with your assessment here except...

> So the issue is specific to wide and short dataframes (which are rare)

One gets a wide, short DataFrame when using `axis=1` on a long narrow DataFrame with certain ops. In my personal experience, both wide and short DataFrames as well as using `axis=1` are quite rare, but I don't know if that transfers to most users.

jbrockmendel commented on pandas/core/arrays/boolean.py:
```diff
@@ -404,3 +405,25 @@ def _accumulate(
+        **kwargs,
+    ):
+        arr = cast(BaseMaskedArray, self.astype("Int8"))
+        return arr.sum(skipna=skipna, min_count=min_count, axis=axis, **kwargs)
```
this seems unrelated.  why is the parent class method not enough here?

topper-123 replied:
Yes, you are right, this isn't needed. I've pushed a new version.

jbrockmendel commented:
I'd like to give this another once-over, but am officially declaring that im OK with the perf impact

topper-123 commented:
> > So the issue is specific to wide and short dataframes (which are rare)
> 
> One gets a wide, short DataFrame when using `axis=1` on a long narrow DataFrame with certain ops. In my personal experience, both wide and short DataFrames as well as using `axis=1` are quite rare, but I don't know if that is also the experience of users.

Yeah, I agree. I made an attempt to address that (`when aggregating over axis=1 (much rarer than aggregation over axis=0, but is already slow from the need to transform`), though I meant transpose, not transform.

topper-123 committed:
remove special BooleanArray.sum method

jbrockmendel commented on pandas/core/arrays/boolean.py:
```diff
@@ -404,3 +405,14 @@ def _accumulate(
+        **kwargs,
+    ):
+        arr = cast(BaseMaskedArray, self.astype("Int8"))
+        return arr.prod(skipna=skipna, min_count=min_count, axis=axis, **kwargs)
```
same question as for sum, why is this needed?

topper-123 replied:
Ups, yes. I've removed it. Thanks.

jbrockmendel commented on pandas/core/arrays/categorical.py:
```diff
@@ -2099,6 +2099,10 @@ def _reverse_indexer(self) -> dict[Hashable, npt.NDArray[np.intp]]:
 
+    def _reduce_and_wrap(self, name: str, *, skipna: bool = True, kwargs):
+        result = self._reduce(name, skipna=skipna, **kwargs)
+        return type(self)([result], dtype=self.dtype)
```
do we get here with e.g. any/all?

topper-123 replied:
Categorical doesn't support any/all, IDK why actually, seems like it could, if the categories do. 

Do you have any specific issue in mind?

topper-123 replied:
gentle ping...

jbrockmendel replied:
maybe a comment that if any/all are ever supported then we shouldnt do this wrapping?

topper-123 committed:
remove BooleanArray.prod

topper-123 committed:
fixes

topper-123 committed:
Update doc/source/whatsnew/v2.1.0.rst

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

topper-123 commented on doc/source/whatsnew/v2.1.0.rst:
```diff
@@ -14,10 +14,43 @@ including other versions of pandas.
-enhancement1
-^^^^^^^^^^^^
+Reductions maintain extension dtypes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
That sounds better to me.

topper-123 commented on doc/source/user_guide/integer_na.rst:
```diff
@@ -126,13 +126,20 @@ These dtypes can be merged, reshaped & casted.
+
+    When doing reduction operations (:meth:`~DataFrame.sum` etc.) on numeric-only data
+    frames the integer array dtype will be maintained. Previously, the dtype of reduction
+    result would have been a numpy numeric dtype.
```
I'll wait to see what the consensus is here.

jbrockmendel replied:
I agree with Joris

topper-123 replied:
Ok, I've removed the note.

jorisvandenbossche commented on pandas/tests/arrays/integer/test_reduction.py:
```diff
@@ -0,0 +1,44 @@
+    if op in {"sum", "prod", "min", "max"}:
+        assert isinstance(result, np.int64)
+    else:
+        assert isinstance(result, int)
```
This is never reached?

mroeschke replied:
Agreed this doesn't look reached based on the parameterized ops

jorisvandenbossche commented on pandas/tests/extension/masked_shared.py:
```diff
@@ -64,6 +64,37 @@ def check_reduce(self, ser: pd.Series, op_name: str, skipna: bool):
+        else:
+            raise TypeError("not supposed to reach this")
+
+        result = arr._reduce_and_wrap(op_name, skipna=skipna, kwargs={})
```
Instead of calling this directly, should we test with `result = DataFrame({"col": arr}).op(...).array` instead?

jorisvandenbossche commented on pandas/tests/extension/base/reduce.py:
```diff
@@ -66,6 +67,15 @@ def test_reduce_series(self, data, all_numeric_reductions, skipna):
             self.check_reduce(s, op_name, skipna)
 
+    @pytest.mark.parametrize("skipna", [True, False])
+    def test_reduce_and_wrap(self, data, all_numeric_reductions, skipna):
```
```suggestion
    def test_reduce_frame(self, data, all_numeric_reductions, skipna):
```

(to name it for what it is meant for user-facing, matching the naming of the other series test?)

topper-123 replied:
Yeah, I like that better, I've also renamed `check_reduce_and_wrap` to `check_reduce_frame`.

jorisvandenbossche commented on pandas/tests/extension/decimal/test_decimal.py:
```diff
@@ -115,6 +115,10 @@ def check_reduce(self, s, op_name, skipna):
 
+    @pytest.mark.skip("tests not written yet")
+    def check_reduce_and_wrap(self, ser: pd.Series, op_name: str, skipna: bool):
+        pass
```
I think it would be good to implement at least one of those, to have coverage of the base class (our own EAs currently (mostly?) override the base class method?) 

(and decimal should be a good candidate, for example just testing that the min/max of DataFrame with decimal column still has decimal dtype)

topper-123 replied:
I've added tests for `Decimal`.

jorisvandenbossche commented on pandas/tests/extension/masked_shared.py:
```diff
@@ -64,6 +64,37 @@ def check_reduce(self, ser: pd.Series, op_name: str, skipna: bool):
+        else:
+            raise TypeError("not supposed to reach this")
+
+        result = arr._reduce_and_wrap(op_name, skipna=skipna, kwargs={})
```
Instead of calling this directly, should we test with `result = DataFrame({"col": arr}).op(...).array` instead?

topper-123 replied:
Yes, that would be good too. I've added it, but I think because `__reduce_and_wrap` is part of the `ExtensionArray` interface, it't be good to test that also, so I test both in the updated version.

jorisvandenbossche reviewed (commented):
Added a bunch of minor comments / question, but generally looks good to me! Approach looks fine, it's not always the cleanest, but I agree that is not always avoidable.

jorisvandenbossche commented:
The top post mentions "https://github.com/pandas-dev/pandas/issues/51446 (the last one intended, not done yet)". Has that changed in the meantime? (just to ensure you update the top post to close/not close that issue)

topper-123 committed:
Update pandas/core/array_algos/masked_reductions.py

Co-authored-by: Joris Van den Bossche <jorisvandenbossche@gmail.com>

topper-123 commented on pandas/core/array_algos/masked_reductions.py:
```diff
@@ -119,13 +119,13 @@ def _minmax(
-        else:
+            return func(values, where=~mask, axis=axis, initial=subset[0])
             # min/max with empty array raise in numpy, pandas returns NA
+        else:
```
yes.

topper-123 replied:
That could well be right. Let me look into it.

topper-123 commented on pandas/core/arrays/masked.py:
```diff
@@ -1088,25 +1092,62 @@ def _reduce(self, name: str, *, skipna: bool = True, **kwargs):
+        )
+
+    def _reduce_and_wrap(self, name: str, *, skipna: bool = True, kwargs):
+        df = self.reshape(-1, 1)
```
Hmm, that could well be right. Let me look into it and I'll get back.

topper-123 replied:
The comment wrapping in github is started to get to the point where I' sure if I've answered everything, But I've made a new version that uses 1d arrays like suggested here, it works the same and is simpler because we avoid possibly complexity of 2d arrays. There is still some 2d code improvements is this PR, but IMO we should keep it in because it makes the the 2d ops better.

topper-123 commented on pandas/core/arrays/masked.py:
```diff
@@ -1192,23 +1234,29 @@ def std(
+        )
 
-    def any(self, *, skipna: bool = True, **kwargs):
+    def any(self, *, skipna: bool = True, axis: AxisInt | None = 0, **kwargs):
```
I'll look into it, coukd be connected to your previous comment.

topper-123 replied:
I think this works, and I've made another version, will if it passes, and then I'll look into your other comments

jreback commented on pandas/core/arrays/arrow/array.py:
```diff
@@ -1533,6 +1533,12 @@ def _reduce(self, name: str, *, skipna: bool = True, **kwargs):
 
         return result.as_py()
 
+    def _reduce_and_wrap(self, name: str, *, skipna: bool = True, kwargs):
```
why are you adding another method here? what's wrong with just fixing _reduce?

topper-123 replied:
`_reduce` on 1d arrays only returns a scalar and we can't differentiate between scalars from reductions from e.g. numpy.int64 and pandas.Int64() arrays. Reductions that return `pd.NA` are just as bad, because `pd.NA` holds no dtype info.

Also, we can't supply `keepdims` to `_reduce`, because pandas raises when `keepdims` is given as a parameter in the reduction methods.

jreback replied:
so why don't u just update _reduce?

jbrockmendel replied:
There was a version similar to this that added a keepdims kwd to _reduce and we decided that this was better bc it didn't require a deprecation path for 3rd party EAs

topper-123 replied:
_reduce calls other methods, e.g. sum. It's in those methods the failures happen when we give keepdims=True and those methods are public. Do we want to change their signatures (and the ._reduce signature) to include keepdims?

topper-123 replied:
If it's just ading the keepdims keyword to `_reduce`, that will be relatively technically easy. It's adding it to `sum` etc. that probably will take more effort. Also note @jbrockmendel comment about deprecation path.

jorisvandenbossche replied:
Compatibility concerns aside, I think the `keepdims` argument to `_reduce` is the nicer solution. 
But, external EAs don't have this keyword, so that means we would need to add some compatibility code anyway, everywhere we call the `_reduce` method (check if it supports the new keyword, and if not still wrap the result in `np.array([res])` array, just like the current base implementation of `_reduce_and_wrap` does). With that, I am not sure that will be an improvement over the current solution.

jbrockmendel replied:
If a hypothetical EA wanted to do a reduction lazily, that would be much easier with a keepdims keyword than with a _reduce_and_wrap method.  Just a thought, not worth contorting ourselves over a hypothetical EA

topper-123 replied:
I'm not opposed to a `keepdims` parameter in `_reduce`, except for the compatibility concerns, but I would like a decision.

One way to address the compatibility concerns could be to introspect the signature of `_reduce` to see if it has a `keepdims` parameter or not. If it does, call `_reduce` with `keepdims=True` when doing dataframe reductions. If it doesn't, call it without a `keepdims` parameter and emit a warning that `keepdims` will become required in the future.

topper-123 commented on pandas/tests/arrays/integer/test_reduction.py:
```diff
@@ -0,0 +1,44 @@
+
+@pytest.mark.parametrize("op", ["sum", "min", "max", "prod"])
+def test_preserve_dtypes(op):
+    # TODO(#22346): preserve Int64 dtype
+    # for ops that enable (mean would actually work here
+    # but generally it is a float return value)
```
This was just moved from another test.  For 1d arrays, reduction. ops won't preserve Int64 dtype, so the comment should be removed, unless we implement scalar Int64 (which we won't AFAIK).

I think this test could be expanded though to treat the other reduction also.

topper-123 committed:
small cleanup

topper-123 committed:
small cleanup

topper-123 committed:
Merge branch 'master' into reduction_dtypes_II

topper-123 committed:
Merge branch 'master' into reduction_dtypes_II

topper-123 committed:
only reduce 1d

topper-123 committed:
Merge branch 'reduction_dtypes_II' of https://github.com/topper-123/pandas into reduction_dtypes_II

topper-123 committed:
fix after #53418

topper-123 commented:
This fails currently, but will pass again after #53482 is merged.

topper-123 committed:
Merge branch 'master' into reduction_dtypes_II

topper-123 committed:
update according to comments

topper-123 committed:
revome note

topper-123 commented:
I think I've addressed all comments, but a decision is needed about doing this as a separate method, or as a key word in `_reduce`. That discussion is [here](https://github.com/pandas-dev/pandas/pull/52788#discussion_r1209527421).

topper-123 commented on pandas/core/array_algos/masked_reductions.py:
```diff
@@ -119,13 +119,13 @@ def _minmax(
         if subset.size:
-            return func(subset)
-        else:
+            return func(values, where=~mask, axis=axis, initial=subset[0])
```
I've changed back to use `subset`.  To use `where=~mask` the `func` requires `initial` to have a value, so there is no real saving by using `where=~mask` in this case.

topper-123 committed:
update _minmax

topper-123 committed:
Merge branch 'master' into reduction_dtypes_II

jorisvandenbossche commented on pandas/core/array_algos/masked_reductions.py:
```diff
@@ -119,11 +119,11 @@ def _minmax(
         subset = values[~mask]
         if subset.size:
-            return func(subset)
+            return func(subset, axis=axis)
```
Note that now you are using `subset` again on this line, passing this `axis` is not doing anything (and would actually raise an error if you would pass `axis=1` here)

(it doesn't really matter in practice because we never call this with an axis=1, but seeing `axis` passed through might give the false impression that this algo actually supports 2D data, while that is not the case)

jbrockmendel replied:
@topper-123 can you address this?

topper-123 replied:
Oh, I thought I had answered this, apparently not...

`func` here is either `np.min` or `np.max`, so supplying `axis=axis` will not raise here, but will work as expected AFAIKS.

Additionally, without the `axis=axis` part, `func(subset)` is similar to `np.max|min(subset, axis=None)`. Not a problem for 1d arrays, but will be a problem if we ever want to support `df.min(axis=None)` using 2d masked arrays. (I'm not sure we want to support 2d masked arrays or are going all in on arrow?)

topper-123 committed:
Merge branch 'master' into reduction_dtypes_II

topper-123 committed:
Merge branch 'master' into reduction_dtypes_II

topper-123 commented:
Ping. I think I need a decision if the `_reduce_and_wrap` way is the way forward or not.

topper-123 committed:
Merge branch 'master' into reduction_dtypes_II

topper-123 committed:
Merge branch 'master' into reduction_dtypes_II

topper-123 committed:
Merge branch 'master' into reduction_dtypes_II

topper-123 committed:
Merge branch 'master' into reduction_dtypes_II

topper-123 committed:
REF: add keepdims parameter to ExtensionArray._reduce + remove ExtensionArray._reduce_and_wrap

topper-123 committed:
REF: add keepdims parameter to ExtensionArray._reduce + remove ExtensionArray._reduce_and_wrap

topper-123 committed:
fix whatsnew

jbrockmendel commented:
I haven't re-reviewed in a while, but I was happy here as of the last time I looked at it.  If Joris is on board, let's do it.

topper-123 commented:
I've actually just this morning made a new version using `_reduce` only (i.e. scrapping `_reduce_and_wrap`)`. I prefer this new version (because we now only have one reduction methods instead of two), but if the other is preferred, it is easy to revert back.

I maintain backward compat in this new version by:
1. making the `keepdim` parameter keyword only
2. only calling it with `keepdims=True` if the `_reduce` method signature has a parameter named "keepdims"
3. if the `_reduce` method signature does not have a parameter named "keepdims", `_reduce` gets called without supplying the `keepdims` parameter, we emit a `FutureWarning` and take care of wrapping the (scalar) reduction result in a ndarray before passing it on.

This is possible because `_reduce_and_wrap` was actually only called inside the `blk_func` inside `DataFrame._reduce`, so by doing some introspection there we can keep backward compat. See new version:

https://github.com/pandas-dev/pandas/blob/f85deab32d337775f6552d092f6bf3054fb60b88/pandas/core/frame.py#L10874-L10895.

Notice especially the FutureWarning starting on line 10885. This will allow us to not require `keepdims` now, even though `keepdims` is in the signature of `ExtensionArray._reduce`.  In v3.0, we will drop the signature checking and only call `values._reduce` with `keepdims=True`, i.e. it will fail without a `keepdims` parameter in v3.0.

Check out the `test_reduction_without_keepdims` test in `pandas/tests/extension/decimal/test_decimal.py` for a test of what happens when extensionarrays don't have a keepdim parameter in their `_reduce` method.

Thoughts? I prefer this new version, but it's is easy to revert back if needed.

topper-123 committed:
fix _reduce call

topper-123 committed:
Merge branch 'master' into reduction_dtypes_II

topper-123 commented:
Ping...

mroeschke reviewed (commented):
Looks pretty good. Just a single comment from me https://github.com/pandas-dev/pandas/pull/52788/files#r1261463068

topper-123 committed:
Merge branch 'master' into reduction_dtypes_II

topper-123 committed:
Merge branch 'master' into reduction_dtypes_II

topper-123 committed:
simplify test

topper-123 committed:
add tests for any/all

topper-123 commented:
I've updated, added  some  + fixed some dtype issues with `DataFrame.any/all`.

mroeschke commented on pandas/tests/arrays/integer/test_reduction.py:
```diff
@@ -0,0 +1,120 @@
+        ["median", Series([2, 2], index=["B", "C"], dtype="Float64")],
+        ["var", Series([2, 2], index=["B", "C"], dtype="Float64")],
+        ["std", Series([2**0.5, 2**0.5], index=["B", "C"], dtype="Float64")],
+        ["skew", Series([pd.NA, pd.NA], index=["B", "C"], dtype="Float64")],
```
In a follow up if you could add `kurt` (if supported) that would be great

mroeschke commented:
Thanks this is an awesome improvement @topper-123

jbrockmendel commented:
nice! thanks for sticking with this @topper-123

topper-123 commented:
Yes, but worth the slow process, I wouldn't have thought of this way to add a parameter to `_reduce` in a backward compatible way if this had gone faster 😄.
